src/experiments/title-generation/components/TitleToolbar.tsx

```diff
@@ -0,0 +1,226 @@
+									)
+								);
+							}}
+							__nextHasNoMarginBottom
+						/>
+						<Button
+							variant="secondary"
+							style={{ marginTop: '15px' }}
+							onClick={() => onSelect(title, i)}
+						>
+							{__('Select', 'ai')}
+						</Button>
+					</FlexItem>
+				);
+			})}
+		</Flex>
+	);
+}
+
+/**
+ * Generates titles for the given post ID and content.
+ *
+ * @param {number} postId  The ID of the post to generate a title for.
+ * @param {string} content The content of the post to generate a title for.
+ * @return {Promise<string[]>} A promise that resolves to the generated titles.
+ */
+async function generateTitles(
+	postId: number,
+	content: string
+): Promise<string[]> {
+	return executeAbility('ai/title-generation', {
+		content,
+		post_id: postId,
+	})
+		.then((response) => {
+			if (
+				response &&
+				typeof response === 'object' &&
+				'titles' in response
+			) {
+				return response.titles as string[];
+			}
+
+			return [];
+		})
+		.catch((error) => {
+			throw new Error(`Error generating titles: ${error.message}`);
+		});
+}
+
+/**
+ * TitleToolbar component.
+ *
+ * Provides Generate/Re-generate button.
+ *
+ * @return {JSX.Element} The toolbar component.
+ */
+export default function TitleToolbar(): JSX.Element | null {
+	const postId = select('core/editor').getCurrentPostId();
+	const postType = select('core/editor').getCurrentPostType();
+	const content = select('core/editor').getEditedPostContent();
+	const title = select('core/editor').getEditedPostAttribute('title');
+
+	const { editPost } = useDispatch('core/editor');
+
+	const [isGenerating, setIsGenerating] = useState(false);
+	const [isOpen, setOpen] = useState(false);
+	const [data, setData] = useState<string[]>([]);
+
+	const openModal = () => setOpen(true);
+	const closeModal = () => {
+		setOpen(false);
+		setData([]);
+	};
+
+	const hasTitle = title.trim().length > 0;
+	const buttonLabel = hasTitle
+		? __('Re-generate', 'ai')
+		: __('Generate', 'ai');
+
+	/**
+	 * Handles the generate/re-generate button click.
+	 */
+	const handleGenerate = async () => {
+		setIsGenerating(true);
+		(dispatch('core/notices') as any).removeNotice(
+			'ai_title_generation_error'
+		);
+
+		try {
+			const generatedTitles = await generateTitles(postId, content);
+			setData(generatedTitles);
+			openModal();
+		} catch (error: any) {
+			(dispatch('core/notices') as any).createErrorNotice(error, {
+				id: 'ai_title_generation_error',
+				isDismissible: true,
+			});
+			setData([]);
+		} finally {
+			setIsGenerating(false);
+		}
+	};
+
+	/**
+	 * Handles selecting a title.
+	 *
+	 * @param {string} selectedTitle The selected title.
+	 */
+	const handleSelectTitle = async (selectedTitle: string) => {
+		const isDirty = select('core/editor').isEditedPostDirty();
+		editPost({
+			title: selectedTitle,
+		});
+		closeModal();
+		if (!isDirty) {
+			await (dispatch('core') as any).saveEditedEntityRecord(
+				'postType',
+				postType,
+				postId
+			);
+		}
```

Comment:
I don't think we should be autosaving. In my testing, I meant to change the title, then go back to the posts list without saving. I was really confused as to why the title had changed when I didn't save. Took me a moment to figure out where was no way to change the title this way without it saving.

Reply:
Removed in https://github.com/WordPress/ai/pull/83/commits/0390cd4f81a42c767cbadd583691180daff2e9fe

Reply:
This isn't saving a revision, though, it's saving the whole post. A revision would be fine.